### DISCUSSION TITLE
Online Groove Pattern Library

### DISCUSSION THREAD

**Discussion by Olli** [hydrogen-music/hydrogen#1059] [Ideas]
A look on other commercial drum software shows, that nearly every software has a groove pattern library included or online available. 
So maybe this is also possible for hydrogen? I know there are patterns available which one can find around the internet.
But can imagine that it would be very nice and convenient to have such a pattern library integrated in Hydrogen similar to the Drumkit library. 
I would be very convenient when it's possible to get for example a nice Reggae groove with 2 variations, a intro, a filling and a ending in no time only for the little arrangement made in a rainy Sunday afternoon (without searching the internet and getting distracted) ;)
This would also imply that there are conventions how an drumkit must be build (in terms of what sound comes out when what note is triggered, how many layers are for minimum). 

What do you think of it?

**Comment by theGreatWhiteShark** [comment]
Hey @Olli ,

In practical terms what's a show stopper here is the limitation of the current drumkit format since it does not feature (enough) metadata. When programming a pattern with one drumkit and changing to another all notes of the first instrument of the drumkit will be mapped to the first one in the other drumkit. This way your e.g. snare notes can end up being interpreted as a tom.

With respect to further development: I'm not speaking on behalf of the whole developing team but AFAICS having a dedicated pattern library in addition to a nice window in Hydrogen and a server providing the patterns is possible but will most probably take a long, long time. Not so much because of the work required here but because of all the other things we already need/promised to do as well. After all, Hydrogen is not backed by a company but maintained/tweaked by a few people in the spare time.

However, it's a feature frequently requested and would ease handling Hydrogen (especially for first-day users). What I can think of is a mixed approach with the devs fixing the switching-between-drumkits-issue and the community handling the sharing of patterns. E.g. there could be dedicated thread(s) here in the *Discussions* sections where people can post their patterns (and tag them for better searchability). Or someone from the community could volunteer in moderating/sorting the pattern in threads (or a dedicated github repo). Either way, I think when sharing the effort we can enable pattern exchange in the Hydrogen user community.

**Comment by Olli** [comment]
Yes that's what I expected - the mapping. Well can't be established a convention about how drumkits must be build? This also may have good implications to the quality of the drumkits. 

I see the point, that it make much much work. When looking from outside it seems be not many work but I've been looking into the code and how QT draws the interface and how the interface is been build up. 
TBH it's way much more work than I thought.

**Comment by mauser** [comment]
Hi @Olli,

there is already the "general midi" convention (a "standard" coming from the MIDI area, take a look [here](http://computermusicresource.com/GM.Percussion.KeyMap.html) for the percussion part). The problem is that it is hard to agree on a good standard if you want to keep all people happy. For example if you agree on the GM standard, each kit has two (different) bassdrums, but only two crash cymbals (+ splash and chinese). Some people want to have 8 toms, others just 2.

What @theGreatWhiteShark meant with metadata is that it could be promising to tag an instrument with some sort of concept, like "Snare 1". Those tags could be given *after* a drumkit has been constructed and so a drumkit could be mapped to a pattern (very shortly described :smiley: ).

**Comment by Olli** [comment]
Ok i got it. I did take a look into the drumkit xml ... there is no hint what note is assigned to a instrument (I hope I don't miss something). 
And the Note-Sound mapping can't be GM-like because GM seem to be too limited for this purpose.
Did I understand it correctly?

**Comment by theGreatWhiteShark** [comment]
Yes, that's right.

Also tagging the existing drumkits and crafting an elegant conversion function will take up some time. After all, if a pattern contains e.g. a snare rimshot but your drumkit does not it would most probably better to replace it by a snare hit than omitting it.

**Comment by trebmuh** [comment]
Just adding quickly to the discussion (in case of @Olli missed it or for the purpose of enlightening potential readers) that we have a [song & pattern repository](https://github.com/hydrogen-music/Song-and-pattern-repository) which although far from being perfect, already provide with a few different pattern linked for different drumkits.
